components/authz-service/server/v2/rules_property_test.go

```diff
@@ -18,17 +21,511 @@ import (
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+
+				_, err = cl.DeleteRule(ctx, &api.DeleteRuleReq{Id: reqs.rules[0].Id})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				// Can still find a staged-deleted rule if the rule was previously applied!
+				rDeleted, err := cl.GetRule(ctx, &api.GetRuleReq{Id: reqs.rules[0].Id})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+
+				_, err = cl.GetRule(ctx, &api.GetRuleReq{Id: reqs.rules[0].Id})
+				deletedWhenApplied := err != nil && strings.Contains(err.Error(), "could not find")
+
+				return deletedWhenApplied &&
+					ruleMatches(reqs.rules[0], *rDeleted.Rule) &&
+					rDeleted.Rule.Status == "applied" && // TODO: wrong!
+					!rDeleted.Rule.Deleted // TODO: wrong!
+			},
+			createProjectAndRuleGen,
+		))
+
+	properties.Property(`deletes an applied and staged rule and applies deletion
+confirming rule found but marked for deletion while deletion staged
+but rule no longer found once deletion is applied`,
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				_, err = cl.CreateRule(ctx, &reqs.rules[0])
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+
+				updateReq := api.UpdateRuleReq{
+					Id:         reqs.rules[0].Id,
+					ProjectId:  reqs.rules[0].ProjectId,
+					Name:       reqs.rules[0].Name + " updated",
+					Type:       reqs.rules[0].Type,
+					Conditions: reqs.rules[0].Conditions,
+				}
+				_, err = cl.UpdateRule(ctx, &updateReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				_, err = cl.DeleteRule(ctx, &api.DeleteRuleReq{Id: reqs.rules[0].Id})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				_, err = cl.GetRule(ctx, &api.GetRuleReq{Id: reqs.rules[0].Id})
+				markedforDeletion := err != nil && strings.Contains(err.Error(), "marked for deletion")
+
+				cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+
+				_, err = cl.GetRule(ctx, &api.GetRuleReq{Id: reqs.rules[0].Id})
+				fullyDeleted := err != nil && strings.Contains(err.Error(), "could not find")
+
+				return markedforDeletion && fullyDeleted
+			},
+			createProjectAndRuleGen,
+		))
+
+	properties.TestingRun(t)
+}
+
+func getGopterParams(seed int64) *gopter.Properties {
+	params := gopter.DefaultTestParametersWithSeed(seed)
+	params.MinSize = 1             // otherwise, we'd get zero-length "conditions" slices
+	params.MaxSize = 25            // otherwise, we may exceed GRPC capacity in ListRules
+	params.MinSuccessfulTests = 25 // reduced from default 100 to avoid timeouts in CI!
```

Comment:
I guess we could expand the timeouts if this is insufficient for us.

Assuming this fails because there's no output for quite a while... I wonder if "output a `"."` for every 10 runs" would be another worthy upstream issue. 🤔